master/buildbot/www/authz/authz.py

```diff
@@ -89,10 +89,13 @@ def assertUserAllowed(self, ep, action, options, userDetails):
                     if owner is not None:
                         roles.update(
                             self.getOwnerRolesFromUser(userDetails, owner))
-                if rule.role not in roles:
-                    if rule.defaultDeny:
-                        raise Forbidden(
-                            "you need to have role '%s'" % (rule.role,))
-                elif not rule.defaultDeny:
-                    defer.returnValue(None)
+                for role in roles:
+                    if self.match(role, rule.role):
```

Comment:
Testing regexmatcher vs regex matcher is not very relevant for me. We dont need to test all the combinations, we can just test that the regex matcher is working as expected, and then only test with fnmatchstrmatcher (or a fake matcher).

As for testing the appropriate usecase, I agree on solution 2. If we need to have a (or set of) test with a specific rule combination, we shall implement it like this.